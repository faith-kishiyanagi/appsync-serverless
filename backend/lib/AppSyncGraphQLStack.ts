import * as cdk from '@aws-cdk/core';
import * as appsync from '@aws-cdk/aws-appsync';
import * as lambda from '@aws-cdk/aws-lambda';

export class AppSyncGraphQLStack extends cdk.Stack {

  readonly api: appsync.GraphqlApi;

  constructor(scope: cdk.Construct, id: string, resolver: lambda.Function, props?: cdk.StackProps) {
    super(scope, id, props);

    this.api = new appsync.GraphqlApi(this, 'GqlApi', {
      name: 'blog-graphql-api',
      schema: appsync.Schema.fromAsset('graphql/schema.graphql'),
      authorizationConfig: {
        defaultAuthorization: {
          authorizationType: appsync.AuthorizationType.API_KEY,
          apiKeyConfig: {
            expires: cdk.Expiration.after(cdk.Duration.days(2)),
          }
        }
      }
    })

    const postResolver = this.api.addLambdaDataSource('PostResolver', resolver)

    postResolver.createResolver({
      typeName: 'Query',
      fieldName: 'allPosts'
    });
    postResolver.createResolver({
      typeName: 'Query',
      fieldName: 'findOnePostBy'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'createPost'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'updatePost'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'deletePost'
    });

    new cdk.CfnOutput(this, 'AppSyncUrl', {
      value: this.api.graphqlUrl
    });
    new cdk.CfnOutput(this, 'AppSyncApiKey', {
      value: this.api.apiKey || ''
    });
    new cdk.CfnOutput(this, 'ProjectRegion', {
      value: this.region
    });
  }
}
