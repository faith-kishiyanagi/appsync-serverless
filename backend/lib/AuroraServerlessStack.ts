import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as rds from '@aws-cdk/aws-rds';
import { AuroraCapacityUnit } from '@aws-cdk/aws-rds';

export class AuroraServerlessStack extends cdk.Stack {

  readonly vpc: ec2.Vpc;
  readonly blogDB: rds.ServerlessCluster;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.vpc = new ec2.Vpc(this, 'BlogAppVpc', {
      cidr: '10.0.1.0/24',
      natGateways: 0,
      subnetConfiguration: [
        {cidrMask: 26, name: 'Isolated', subnetType: ec2.SubnetType.ISOLATED}
      ],
    })

    this.blogDB = new rds.ServerlessCluster(this, 'BlogDbAuroraServerlessCluster', {
      engine: rds.DatabaseClusterEngine.AURORA_MYSQL,
      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'BlogDbParameterGroup', 'default.aurora-mysql5.7'),
      defaultDatabaseName: 'BlogDB',
      vpc: this.vpc,
      vpcSubnets: {subnetType: ec2.SubnetType.ISOLATED},
      scaling: {
        autoPause: cdk.Duration.minutes(5),
        minCapacity: AuroraCapacityUnit.ACU_1,
        maxCapacity: AuroraCapacityUnit.ACU_2,
      }
    })
  }
}
