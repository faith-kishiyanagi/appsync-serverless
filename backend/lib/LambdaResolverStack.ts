import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as rds from '@aws-cdk/aws-rds';

export class LambdaResolverStack extends cdk.Stack {

  readonly postFunction: lambda.Function;

  constructor(scope: cdk.Construct, id: string, db: rds.ServerlessCluster, props?: cdk.StackProps) {
    super(scope, id, props);

    this.postFunction = new lambda.Function(this, 'PostFunction', {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: new lambda.AssetCode('functions/post'),
      handler: 'src/index.handler',
      memorySize: 256,
      timeout: cdk.Duration.seconds(30),
      environment: {
        CLUSTER_ARN: db.clusterArn,
        SECRET_ARN: db.secret?.secretArn || '',
        DB_NAME: 'BlogDB',
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1'
      }
    })

    db.grantDataApiAccess(this.postFunction);
  }
}
