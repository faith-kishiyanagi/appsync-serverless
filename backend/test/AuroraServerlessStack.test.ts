import '@aws-cdk/assert/jest'
import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { AuroraServerlessStack } from '../lib/AuroraServerlessStack';

const app = new cdk.App();
const stack = new AuroraServerlessStack(app, 'AuroraServerless');

test('Stack snapshot', () => {
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('VPC', () => {
  expect(stack).toHaveResource('AWS::EC2::VPC');
  expect(stack).toCountResources('AWS::EC2::Subnet', 2)
});

test('Aurora Serverless', () => {
  expect(stack).toHaveResource('AWS::RDS::DBCluster', {
    ScalingConfiguration: {
      MinCapacity: 1,
      MaxCapacity: 2,
      AutoPause: true,
      SecondsUntilAutoPause: 300,
    }
  });
});
