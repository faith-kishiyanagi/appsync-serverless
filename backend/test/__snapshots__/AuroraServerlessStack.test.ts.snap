// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack snapshot 1`] = `
Object {
  "Resources": Object {
    "BlogAppVpcA4BA08F3": Object {
      "Properties": Object {
        "CidrBlock": "10.0.1.0/24",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "AuroraServerless/BlogAppVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "BlogAppVpcIsolatedSubnet1RouteTable65DB2CE1": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "AuroraServerless/BlogAppVpc/IsolatedSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "BlogAppVpcA4BA08F3",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "BlogAppVpcIsolatedSubnet1RouteTableAssociation29680609": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "BlogAppVpcIsolatedSubnet1RouteTable65DB2CE1",
        },
        "SubnetId": Object {
          "Ref": "BlogAppVpcIsolatedSubnet1Subnet6E8FF111",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "BlogAppVpcIsolatedSubnet1Subnet6E8FF111": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "AuroraServerless/BlogAppVpc/IsolatedSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "BlogAppVpcA4BA08F3",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "BlogAppVpcIsolatedSubnet2RouteTable1D112A1E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "AuroraServerless/BlogAppVpc/IsolatedSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "BlogAppVpcA4BA08F3",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "BlogAppVpcIsolatedSubnet2RouteTableAssociationFAD6D859": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "BlogAppVpcIsolatedSubnet2RouteTable1D112A1E",
        },
        "SubnetId": Object {
          "Ref": "BlogAppVpcIsolatedSubnet2Subnet12CC2591",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "BlogAppVpcIsolatedSubnet2Subnet12CC2591": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.64/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "AuroraServerless/BlogAppVpc/IsolatedSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "BlogAppVpcA4BA08F3",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "BlogDbAuroraServerlessClusterB47832F5": Object {
      "DeletionPolicy": "Snapshot",
      "Properties": Object {
        "DBClusterParameterGroupName": "default.aurora-mysql5.7",
        "DBSubnetGroupName": Object {
          "Ref": "BlogDbAuroraServerlessClusterSubnetsAB2728BD",
        },
        "DatabaseName": "BlogDB",
        "Engine": "aurora-mysql",
        "EngineMode": "serverless",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "BlogDbAuroraServerlessClusterSecretCDD0164F",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "BlogDbAuroraServerlessClusterSecretCDD0164F",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "ScalingConfiguration": Object {
          "AutoPause": true,
          "MaxCapacity": 2,
          "MinCapacity": 1,
          "SecondsUntilAutoPause": 300,
        },
        "StorageEncrypted": true,
        "VpcSecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "BlogDbAuroraServerlessClusterSecurityGroup9C6808DE",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBCluster",
      "UpdateReplacePolicy": "Snapshot",
    },
    "BlogDbAuroraServerlessClusterSecretAttachment26906F9D": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "BlogDbAuroraServerlessClusterSecretCDD0164F",
        },
        "TargetId": Object {
          "Ref": "BlogDbAuroraServerlessClusterB47832F5",
        },
        "TargetType": "AWS::RDS::DBCluster",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "BlogDbAuroraServerlessClusterSecretCDD0164F": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Generated by the CDK for stack: ",
              Object {
                "Ref": "AWS::StackName",
              },
            ],
          ],
        },
        "GenerateSecretString": Object {
          "ExcludeCharacters": " %+~\`#$&*()|[]{}:;<>?!'/@\\"\\\\",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{\\"username\\":\\"admin\\"}",
        },
      },
      "Type": "AWS::SecretsManager::Secret",
    },
    "BlogDbAuroraServerlessClusterSecurityGroup9C6808DE": Object {
      "Properties": Object {
        "GroupDescription": "RDS security group",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "BlogAppVpcA4BA08F3",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "BlogDbAuroraServerlessClusterSubnetsAB2728BD": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnets for BlogDbAuroraServerlessCluster database",
        "SubnetIds": Array [
          Object {
            "Ref": "BlogAppVpcIsolatedSubnet1Subnet6E8FF111",
          },
          Object {
            "Ref": "BlogAppVpcIsolatedSubnet2Subnet12CC2591",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
  },
}
`;
